turbopack/crates/turbopack-core/src/module_graph/merged_modules.rs

```diff
@@ -0,0 +1,584 @@
+        let mut lists = vec![];
+        let mut lists_reverse_indices: FxIndexMap<
+            ResolvedVc<Box<dyn MergeableModule>>,
+            FxIndexSet<ListOccurence>,
+        > = FxIndexMap::default();
+
+        // A map of all references betwee modules with the same bitmap. These are all references,
+        // including reexecution edges and cycles. Used to expose additional modules if the
+        // bitmap-groups are split up further.
+        #[allow(clippy::type_complexity)]
+        let mut intra_group_references: FxIndexMap<
+            ResolvedVc<Box<dyn Module>>,
+            FxIndexSet<ResolvedVc<Box<dyn Module>>>,
+        > = FxIndexMap::default();
+        // A map of all references betwee modules with the same bitmap. These are only the
+        // references relevant for execution (ignoring cycles), to find the entries of a group.
+        #[allow(clippy::type_complexity)]
+        let mut intra_group_references_rev: FxIndexMap<
+            ResolvedVc<Box<dyn Module>>,
+            FxIndexSet<ResolvedVc<Box<dyn Module>>>,
+        > = FxIndexMap::default();
+
+        // TODO try to parallelize this loop somehow
+        for chunk_group in &chunk_group_info.chunk_groups {
+            // A partition of all modules in the chunk into several execution traces (orderings),
+            // stored in the top-level lists and referenced here by index.
+            let mut chunk_lists: FxHashMap<RoaringBitmapWrapper, usize> =
+                FxHashMap::with_capacity_and_hasher(
+                    module_merged_groups.len() / chunk_group_info.chunk_groups.len(),
+                    Default::default(),
+                );
+
+            // This is necessary to have the correct order with cycles: a `a -> b -> a` graph would
+            // otherwise be visited as `b->a`, `a->b`, leading to the list `a, b` which is not
+            // execution order.
+            let mut visited = FxHashSet::default();
+
+            module_graph
+                .traverse_edges_from_entries_topological(
+                    chunk_group.entries(),
+                    &mut (),
+                    |parent_info, node, _| {
+                        if parent_info.is_none_or(|(_, r)| r.chunking_type.is_parallel())
+                            && visited.insert(node.module)
+                        {
+                            Ok(GraphTraversalAction::Continue)
+                        } else {
+                            Ok(GraphTraversalAction::Exclude)
+                        }
+                    },
+                    |parent_info, node, _| {
+                        let module = node.module;
+
+                        if let Some(bitmap) = module_merged_groups.get(&module)
+                            && let Some(mergeable_module) =
+                                ResolvedVc::try_downcast::<Box<dyn MergeableModule>>(module)
+                        {
+                            match chunk_lists.entry(bitmap.clone()) {
```

Comment:
could we use references as keys to avoid the `clone`?   the mergable_modules map outlives chunk_lists afaict so it should work out lifetime-wise